lib/charms/grafana_agent/v0/cos_machine.py

```diff
@@ -1,8 +1,166 @@
+
+```python
+    def __init__(
+        self,
+        charm: CharmType,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        metrics_endpoints: Optional[List[dict]] = None,
+        metrics_rules_dir: str = "./src/prometheus_alert_rules",
+        logs_rules_dir: str = "./src/loki_alert_rules",
+        recurse_rules_dirs: bool = False,
+        logs_slots: Optional[List[str]] = None,
+        dashboard_dirs: Optional[List[str]] = None,
+        refresh_events: Optional[List] = None,
+    ):
+```
+
+### Paramenters
+
+- `charm`: The instance of the charm that instantiates `COSMachineProvider`, tipically `self`.
+
+- `relation_name`: If your charmed operator uses a relation name other than `cos-machine` to use
+    the `cos_machine` interface, this is where you have to specify that.
+
+- `metrics_endpoints`: In this parameter you can specify the metrics endpoints that Grafana Agent
+    machine Charmed Operator will scrape.
+
+- `metrics_rules_dir`: The directory in which the Charmed Operator stores its metrics alert rules files.
+
+- `logs_rules_dir`: The directory in which the Charmed Operator stores its logs alert rules files.
+
+- `recurse_rules_dirs`: This paramenters set wheter Grafana Agent machine Charmed Operator has to search
+    alert rules files recursively in the previous two directories or not.
+
+- `logs_slots`: Snap slots to connect to for scraping logs in the form ["snap-name:slot", ...].
+
+- `dashboard_dirs`: List of directories where the dashboards are stored in the Charmed Operator.
+
+- `refresh_events`: List of events on which to refresh relation data.
+
+
+### Example 1 - Minimal instrumentation:
+
+In order to use this object the following should be in the `charm.py` file.
+
+```python
+from charms.grafana_agent.v0.cos_machine import COSMachineProvider
+...
+class TelemetryProviderCharm(CharmBase):
+    def __init__(self, *args):
+        ...
+        self._grafana_agent = COSMachineProvider(self)
+```
+
+### Example 2 - Full instrumentation:
+
+In order to use this object the following should be in the `charm.py` file.
+
+```python
+from charms.grafana_agent.v0.cos_machine import COSMachineProvider
+...
+class TelemetryProviderCharm(CharmBase):
+    def __init__(self, *args):
+        ...
+        self._grafana_agent = COSMachineProvider(
+            self,
+            relation_name="custom-cos-machine",
+            metrics_endpoints=[
+                {"path": "/metrics", "port": 9000},
+                {"path": "/metrics", "port": 9001},
+                {"path": "/metrics", "port": 9002},
+            ],
+            metrics_rules_dir="./src/alert_rules/prometheus",
+            logs_rules_dir="./src/alert_rules/loki",
+            recursive_rules_dir=True,
+            logs_slots=["my-app:slot"],
+            dashboard_dirs=["./src/dashboards_1", "./src/dashboards_2"],
+            refresh_events=["update-status", "upgrade-charm"],
```

Comment:
Odd that the library doesn't already observe `upgrade-charm` on behalf of the charm. We should probably add it in and drop it from the example here.

Reply:
`upgrade-charm` is certainly a bad example here, since [this event will end up triggering a `start` event](https://juju.is/docs/sdk/upgrade-charm-event#heading--emission-sequence) which we observe on behalf of the charm.

Reply:
Assuming that we trust the workflow, `config-changed` is at least a safe default (and also in that emission sequence).

I actually wonder whether we shouldn't make this example smaller. Showing an example with every possible default arg overridden is great for API docs, but the signature and docstrings can help there. This sort of reads like every user _must_ provide all of these fields, when almost none of them are strictly required as long as the defaults are ok.

Reply:
- config-changed is currently observed only if `refresh_events` is None.
- I do not see any observer for `start`

Given the above I think we need to always observe config-changed. And document here that there is no need to observe either config-changed or upgrade-charm.

Reply:
@sed-i sorry I got confused, the `start` event is observer in the machine_charm.py file

Reply:
Let's follow this on [this issue](https://github.com/canonical/grafana-agent-k8s-operator/issues/126) and merge this PR